src/datasource.ts

```diff
@@ -0,0 +1,205 @@
+/*
+** Copyright © 2023 Oracle and/or its affiliates. All rights reserved.
+** Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+*/
+
+import _,{ isString} from 'lodash';
+import { DataSourceInstanceSettings, ScopedVars, MetricFindValue } from '@grafana/data';
+import { DataSourceWithBackend, getTemplateSrv } from '@grafana/runtime';
+import {
+  OCIResourceItem,
+  ResponseParser,
+  //OCIResourceMetadataItem,
+} from './resource.response.parser';
+import {
+  OCIDataSourceOptions,
+  OCIQuery,
+  OCIResourceCall,
+  //QueryPlaceholder,
+  regionsQueryRegex,
+  tenanciesQueryRegex,
+  DEFAULT_TENANCY
+} from "./types";
+//import QueryModel from './query_model';
+
+
+export class OCIDataSource extends DataSourceWithBackend<OCIQuery, OCIDataSourceOptions> {
+  private jsonData: any;
+
+  constructor(instanceSettings: DataSourceInstanceSettings<OCIDataSourceOptions>) {
+    super(instanceSettings);
+    this.jsonData = instanceSettings.jsonData;
+  }
+
+  /**
+   * Filters disabled/hidden queries
+   *
+   * @param {string} query Query
+   */
+  filterQuery(query: OCIQuery): boolean {
+    if (query.hide) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Override to apply template variables
+   *
+   * @param {string} query Query
+   * @param {ScopedVars} scopedVars Scoped variables
+   */
+  applyTemplateVariables(query: OCIQuery, scopedVars: ScopedVars) {
+    const templateSrv = getTemplateSrv();  
+    query.region = templateSrv.replace(query.region, scopedVars);
+    query.tenancy = templateSrv.replace(query.tenancy, scopedVars);
+    if (query.tenancy) {
+      query.tenancy = templateSrv.replace(query.tenancy, scopedVars);
+    }
+    
+    //const queryModel = new QueryModel(query, getTemplateSrv());
+    query.searchQuery = templateSrv.replace(query.searchQuery, scopedVars);
+    return query;
+  }
+
+
+  interpolateProps<T extends Record<string, any>>(object: T, scopedVars: ScopedVars = {}): T {
+    const templateSrv = getTemplateSrv();
+    return Object.entries(object).reduce((acc, [key, value]) => {
+      return {
+        ...acc,
+        [key]: value && isString(value) ? templateSrv.replace(value, scopedVars) : value,
+      };
+    }, {} as T);
+  }
+
+  // // **************************** Template variable helpers ****************************
+
+  // /**
+  //  * Matches the regex from creating template variables and returns options for the corresponding variable.
+  //  * Example:
+  //  * template variable with the query "regions()" will be matched with the regionsQueryRegex and list of available regions will be returned.
+  //  */
+  // metricFindQuery?(query: any, options?: any): Promise<MetricFindValue[]> {
+
+  async metricFindQuery?(query: any, options?: any): Promise<MetricFindValue[]> {
```

Comment:
maybe we can rename to logFindQuery this function

Reply:
I tried that and I think it's a  Grafana function and cannot change it